src/dfx/src/commands/canister/install.rs

```diff
@@ -78,28 +84,44 @@ pub async fn exec(
 
         let canister_id =
             Principal::from_text(canister).or_else(|_| canister_id_store.get(canister))?;
-        let canister_info = CanisterInfo::load(&config, canister, Some(canister_id))?;
-
-        let maybe_path = canister_info.get_output_idl_path();
-        let init_type = maybe_path.and_then(|path| get_candid_init_type(&path));
         let arguments = opts.argument.as_deref();
         let arg_type = opts.argument_type.as_deref();
-        let install_args = blob_from_arguments(arguments, None, arg_type, &init_type)?;
-        let installed_module_hash =
-            read_module_hash(agent, &canister_id_store, &canister_info).await?;
-
-        install_canister(
-            env,
-            agent,
-            &canister_info,
-            &install_args,
-            mode,
-            timeout,
-            call_sender,
-            installed_module_hash,
-            opts.upgrade_unchanged,
-        )
-        .await
+        let canister_info = CanisterInfo::load(&config, canister, Some(canister_id));
+        if let Some(wasm_path) = opts.wasm {
+            // streamlined version, we can ignore most of the environment
+            let install_args = blob_from_arguments(arguments, None, arg_type, &None)?;
+            install_canister_wasm(
+                env,
+                agent,
+                canister_id,
+                canister_info.as_ref().map(|info| info.get_name()).ok(),
+                &install_args,
+                mode,
+                timeout,
+                call_sender,
+                fs::read(wasm_path)?,
```

Comment:
Please add a .with_context() here
```suggestion
                fs::read(wasm_path).with_context(||format!("Unable to read {}", wasm_path.to_string_lossy()))?,
```